### ISSUE TITLE
TOPS-10/20 executable files

### ISSUE THREAD

**Issue by larsbrinkhoff** [larsbrinkhoff/pdp10-its-disassembler#1]
Should I get any spare time, it might be useful to add support for TOPS-10 and TOPS-20 executables.

First, I'd like to know which kind of files there are.

- SAV
- EXE
- SHR?

@rcornwell @b4 @bictorv, can you advise?

**Comment by rcornwell** [comment]
You forgot .LOW and .HGH. Generally .LOW was deleted since it was empty.

**Comment by larsbrinkhoff** [comment]
From KLH10:

```
DEC nonsharable save format:
        <data blocks>
        <entry vector pointer>

    <data block>:
        -<# wds>,,<addr-1>
        <block of that many words>

    <entry vector pointer>:
        <length of vector>,,<addr of vector>

        Vector word 0 is instr to execute to start program.
        Vector word 1 is instr to execute to reenter program.
        Vector word 2 contains program version # info.

        BUT if LH is 254000 then:
                start addr = RH(120)
                reenter addr = RH(124)
                version info in 137
        Actually it appears that the RH may be the start address.
                If it's non-zero, let's use that.

**Comment by larsbrinkhoff** [comment]
From KLH10:

```
DEC sharable SAVE format:
        <directory area>:
                <directory block>
                <entry vector block>
                <optional: PDV (program data vector) block>
                <end block>
        <data area>:
                data pages

Each block has this general format:

                <id code>,,<# words in block, including this word>
                <#-1 remaining words>

    <directory block>:
                1776,,<#>
                <page group descriptors>:
                        <access bits (9)> <27-bit page # in file, 0 if none>
                        <9-bit repeat cnt><27-bit page # in process>

                        Access bits:
                                B1 - pages are sharable
                                B2 - pages are writable
                        Repeat count: # of pages (minus 1) in group.

    <entry vector block>:       ; Optional in TOPS-10
                1775,,3
                <# words in entry vector>
                <addr of entry vector>

    <PDV block>:                ; optional, basically ignore this
                1774,,<#>
                <#-1 words>

    <end block>:
                1777,,1

Entry vector contents are the same as for SAV (non-sharable) format.

TOPS-10 EXEs appear to leave out the entry vector block; in that case,
the contents are taken from
                start addr = RH(120)
                reenter addr = RH(124)
                version info in 137
Again, just as for SAV format.

**Comment by larsbrinkhoff** [comment]
I saw somewhere today that HGH is a unshared high segment, and SHR is a shared high segment.

**Comment by rcornwell** [comment]
.LOW and .HGH were the nonshared low and high segments. .SHR was the high segment for a shared executable. Sometimes there was a non-shared .LOW segment. I am not sure if there was any bit differences between .HGH and .SHR.

**Comment by larsbrinkhoff** [comment]
The high segments loaded by HGH and SHR has base address 400000.

**Comment by larsbrinkhoff** [comment]
I don't see any symbol tables in the descriptions above.

**Comment by atsampson** [comment]
Some TOPS-20 .EXEs I've seen have a separate .SYMBOLS file.

**Comment by atsampson** [comment]
Looking at the symbols files, e.g. [mdl104.symbols](https://github.com/PDP-10/muddle/blob/master/lcml/mdl104.symbols) or teco.symbols/lisp.symbols from Panda, it looks like the first word is an AOBJN pointer to where the symbol table should be in memory, and the rest is a normal list of squoze/value pairs. So dis10 could just have an option to load a separate symbols file after loading the executable.

**Comment by larsbrinkhoff** [comment]
I have been confused about the TOPS-10/20/TENEX executable formats and file types.

I gather there are just two file formats: shareable save file, and compressed/nonsharable save file.  See descriptions from KLH10 above.

File extensions.
- TOPS-10 uses EXE in later versions, LOW, SHR, HGH in older versions.
- TOPS-20 mainly uses EXE.
- TENEX mainly uses SAV. 

Commands.
- TENEX has SAVE for compressed, and SSAVE for sharable.
- TOPS-20 has SAVE for sharable, and CSAVE for compressed.
- TOPS-10 has SAVE... compressed?  Sharable in later versions?

**Comment by larsbrinkhoff** [comment]
Hmm, I was still confused.  I have checked some .SHR files and they are not compressed.  I believe they are just flat raw memory contents from 400000, ending on a (DEC sized) page boundary.  At least that's what the start of the file looks like.  Is there some metadata anywhere?

**Comment by larsbrinkhoff** [comment]
Fixed by #56, #113, #115, #116.
